src/mem_table_flush.rs

```diff
@@ -88,71 +89,105 @@ impl DbInner {
                             }
-                            match flusher.flush_imm_memtables_to_l0().await {
-                                Ok(_) => {
-                                    this.db_stats.immutable_memtable_flushes.inc();
-                                }
-                                Err(err) => {
-                                    error!("error from memtable flush: {err}");
-                                    this.set_error_if_none(err);
-                                }
+                            Err(err) => {
+                                error!("error from memtable flush: {err}");
+                                break Err(err);
                             }
                         }
                     }
                     msg = rx.recv() => {
                         let (rsp_sender, msg) = msg.expect("channel unexpectedly closed");
                         match msg {
                             MemtableFlushThreadMsg::Shutdown => {
-                                is_stopped = true
+                                break Ok(());
                             },
                             MemtableFlushThreadMsg::FlushImmutableMemtables => {
                                 let result = flusher.flush_imm_memtables_to_l0().await;
-                                match &result {
+                                if let Some(rsp_sender) = rsp_sender {
+                                    let res = rsp_sender.send(result.clone());
+                                    if let Err(Err(err)) = res {
+                                        error!("error sending flush response: {err}");
+                                    }
+                                }
+                                match result {
                                     Ok(_) => {
                                         this.db_stats.immutable_memtable_flushes.inc();
                                     }
                                     Err(err) => {
                                         error!("error from memtable flush: {err}");
-                                        this.set_error_if_none(err.clone());
-                                    }
-                                }
-                                if let Some(rsp_sender) = rsp_sender {
-                                    let res = rsp_sender.send(result);
-                                    if let Err(Err(err)) = res {
-                                        error!("error sending flush response: {err}");
-                                        this.set_error_if_none(err);
+                                        break Err(err);
                                     }
                                 }
                             }
                         }
                     }
                 }
+            };
+
+            // respond to any pending msgs
+            let pending_result = map_stopped_task_result(&result);
```

Comment:
how about replace this function with a one-liner like: `let pending_result = result.and_then(|_| Err(BackgroundTaskShutdown))` ?;

it looks easier for me to understand its intention is to transform the `Ok()` result into `Err(BackgroundTaskShutdown)`.